### DISCUSSION TITLE
Static branching within static branching

### DISCUSSION THREAD

**Discussion by alexpghayes** [ropensci/targets#348] [Help] [answered]
I got my previous simulation to work (thank you!) and am now expanding it to include more models. First I statically branch to sample from 2 different models. Now, for each of those two models, I want to fit two different estimators. My first attempt is at https://github.com/alexpghayes/citation-impute-simulations/blob/main/_targets.R but I'm struggling to get the splitting within splitting.

![imagen](https://user-images.githubusercontent.com/15078251/111675917-a88fe900-87eb-11eb-9784-39caa47bce2e.png)

If I use `tidyr::expand(models, estimators)` and a single `tar_map()` call, I end up sampling from each model for each estimator, which is not what I want.

**Comment by alexpghayes** [comment, marked as answer]
So I eventually got the computation graph I wanted in 

![imagen](https://user-images.githubusercontent.com/15078251/111682261-222ad580-87f2-11eb-9e82-5b2131ee5aed.png) 

with some metaprogramming. See https://github.com/alexpghayes/citation-impute-simulations/blob/unholy/_targets.R. But then `targets` errored on me during the make process.

> Error : Target loss_symmetric_svd_model1 tried to branch over svd_model1_symmetric_svd, which is illegal. Patterns must only branch over explicitly declared targets in the pipeline. Stems and patterns are fine, but you cannot branch over branches or global objects. Also, if you branch over a target with format = "file", then that target must also be a pattern.Error: callr subprocess failed: Target loss_symmetric_svd_model1 tried to branch over svd_model1_symmetric_svd, which is illegal. Patterns must only branch over explicitly declared targets in the pipeline. Stems and patterns are fine, but you cannot branch over branches or global objects. Also, if you branch over a target with format = "file", then that target must also be a pattern.

Does this mean I've hit a fundamental limitation of `targets`? Is the solution here to use `tidyr::expand_grid()` to generate the full grid of `models` crossed with `estimators` from the beginning, and to just deal with the fact the duplication of the distributions?

**Reply by wlandau** [reply]
The easiest way I know is a `tar_map()` inside another `tar_map()` (see below). In this case, the inner `tar_map()` is for the `estimate_*` targets. (Alternatively, `filter()` after `expand_grid()` might work but is more prone to mistakes.)

``` r
library(targets)

tar_script({
library(targets)
library(tarchetypes)
tar_map(
  values = list(model = rlang::syms(paste0("model", 1:2))),
  tar_target(
    distribution,
    model(parameters$n, parameters$k, parameters$expected_degree),
    pattern = map(parameters),
    iteration = "list"
  ),
  tar_target(
    realizations,
    purrr::map(1:num_reps, ~sample_sparse(distribution)),
    pattern = map(distribution),
    iteration = "list"
  ),
  tar_target(
    svd,
    svds(distribution),
    pattern = map(population)
  ),
  tar_map(
    values = list(estimator = rlang::syms(c( "full_svds", "symmetric_svd"))),
    tar_target(
      estimate,
      purrr::map(realizations, ~estimator(.x, k = parameters$k)),
      pattern = map(realizations, parameters)
    )
  )
)
})

tar_visnetwork()
```

![](https://i.imgur.com/FfhGqST.png)

<sup>Created on 2021-03-18 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Reply by alexpghayes** [reply]
Do you have an example where this runs? I can construct the right graph this way (https://github.com/alexpghayes/citation-impute-simulations/blob/nested_map/_targets.R), but again `tar_make()` gives a "branching with branch" error.

![imagen](https://user-images.githubusercontent.com/15078251/111699119-b2bee100-8805-11eb-98ae-42716edade17.png)

**Reply by wlandau** [reply]
Did you mean to type `map(distribution)` instead of `map(population)` in the `svd*` targets? https://github.com/alexpghayes/citation-impute-simulations/pull/4 seems to fix the error on my end.

**Reply by alexpghayes** [reply]
Thank you!!

**Reply by liutiming** [reply]
Hi @alexpghayes, the link is broken. I think the others may also benefit from reading your use case if that is possible?

**Reply by liutiming** [reply]
@wlandau the nested tar_map just impressed me again about how extensible targets is 👍

**Reply by alexpghayes** [reply]
@liutiming The link is to an ongoing research project that I prefer not be public just yet. I'll make it public again once I've submitted my manuscript!
